# Damaged Properties

In [1]:
import pandas as pd 
import numpy as np

In [2]:
df = pd.read_csv('../../data/raw/Damaged_Property.csv')

In [3]:
df.head()

CaseID                  Opened                  Closed  \
0  11771122  12/04/2019 04:24:00 PM                     NaN   
1  11770741  12/04/2019 03:14:00 PM                     NaN   
2  11768297  12/04/2019 09:06:00 AM                     NaN   
3  11769772  12/04/2019 12:47:00 PM  12/04/2019 01:30:01 PM   
4  11770411  12/04/2019 02:23:00 PM                     NaN   

                  Updated  Status                        Status Notes  \
0  12/04/2019 05:00:02 PM    Open                            accepted   
1  12/04/2019 05:06:56 PM    Open                                Open   
2  12/04/2019 09:45:02 AM    Open                            accepted   
3  12/04/2019 01:30:01 PM  Closed  meter shop has closed this request   
4  12/04/2019 03:00:01 PM    Open                            accepted   

         Responsible Agency          Category            Request Type  \
0      DPT Meter_Bike Queue  Damaged Property   Damaged Parking_Meter   
1  DPT - Traffic Signal - G  Damaged Property  Damaged Traffic_Signal   
2      DPT Meter_Bike Queue  Damaged Property   Damaged Parking_Meter   
3      DPT Meter_Bike Queue  Damaged Property   Damaged Parking_Meter   
4      DPT Meter_Bike Queue  Damaged Property   Damaged Parking_Meter   

  Request Details                                          Address  \
0   Parking_Meter         Intersection of HAIGHT ST and ASHBURY ST   
1  Traffic_Signal  Intersection of SACRAMENTO ST and MONTGOMERY ST   
2   Parking_Meter     3611 CALIFORNIA ST, SAN FRANCISCO, CA, 94118   
3   Parking_Meter        Intersection of CALIFORNIA ST and POLK ST   
4   Parking_Meter        925 OFARRELL ST, SAN FRANCISCO, CA, 94109   

          Street  Supervisor District                  Neighborhood  \
0      HAIGHT ST                  5.0                Haight Ashbury   
1  SACRAMENTO ST                  3.0            Financial District   
2  CALIFORNIA ST                  2.0  Laurel Heights / Jordan Park   
3  CALIFORNIA ST                  3.0                    Polk Gulch   
4    OFARRELL ST                  6.0                    Tenderloin   

  Police District   Latitude   Longitude                         Point Source  \
0            PARK  37.770008 -122.446930  (37.77000809, -122.44692993)  Phone   
1         CENTRAL  37.793781 -122.402954   (37.79378128, -122.4029541)  Phone   
2        RICHMOND  37.786232 -122.453682  (37.78623199, -122.45368195)  Phone   
3        NORTHERN  37.790577 -122.420685  (37.79057693, -122.42068481)  Phone   
4        NORTHERN  37.784645 -122.420021  (37.78464508, -122.42002106)  Phone   

  Media URL  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN

In [4]:
df.dtypes

CaseID                   int64
Opened                  object
Closed                  object
Updated                 object
Status                  object
Status Notes            object
Responsible Agency      object
Category                object
Request Type            object
Request Details         object
Address                 object
Street                  object
Supervisor District    float64
Neighborhood            object
Police District         object
Latitude               float64
Longitude              float64
Point                   object
Source                  object
Media URL               object
dtype: object

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107762 entries, 0 to 107761
Data columns (total 20 columns):
CaseID                 107762 non-null int64
Opened                 107762 non-null object
Closed                 106980 non-null object
Updated                107762 non-null object
Status                 107762 non-null object
Status Notes           90341 non-null object
Responsible Agency     107762 non-null object
Category               107762 non-null object
Request Type           107762 non-null object
Request Details        107761 non-null object
Address                107762 non-null object
Street                 105920 non-null object
Supervisor District    105920 non-null float64
Neighborhood           105856 non-null object
Police District        105829 non-null object
Latitude               107762 non-null float64
Longitude              107762 non-null float64
Point                  107762 non-null object
Source                 107762 non-null object
Media URL     

## Rename columns

In [6]:
df.rename(columns={'Latitude': 'Y',
                  'Longitude': 'X'}, inplace=True)
df.head()

CaseID                  Opened                  Closed  \
0  11771122  12/04/2019 04:24:00 PM                     NaN   
1  11770741  12/04/2019 03:14:00 PM                     NaN   
2  11768297  12/04/2019 09:06:00 AM                     NaN   
3  11769772  12/04/2019 12:47:00 PM  12/04/2019 01:30:01 PM   
4  11770411  12/04/2019 02:23:00 PM                     NaN   

                  Updated  Status                        Status Notes  \
0  12/04/2019 05:00:02 PM    Open                            accepted   
1  12/04/2019 05:06:56 PM    Open                                Open   
2  12/04/2019 09:45:02 AM    Open                            accepted   
3  12/04/2019 01:30:01 PM  Closed  meter shop has closed this request   
4  12/04/2019 03:00:01 PM    Open                            accepted   

         Responsible Agency          Category            Request Type  \
0      DPT Meter_Bike Queue  Damaged Property   Damaged Parking_Meter   
1  DPT - Traffic Signal - G  Damaged Property  Damaged Traffic_Signal   
2      DPT Meter_Bike Queue  Damaged Property   Damaged Parking_Meter   
3      DPT Meter_Bike Queue  Damaged Property   Damaged Parking_Meter   
4      DPT Meter_Bike Queue  Damaged Property   Damaged Parking_Meter   

  Request Details                                          Address  \
0   Parking_Meter         Intersection of HAIGHT ST and ASHBURY ST   
1  Traffic_Signal  Intersection of SACRAMENTO ST and MONTGOMERY ST   
2   Parking_Meter     3611 CALIFORNIA ST, SAN FRANCISCO, CA, 94118   
3   Parking_Meter        Intersection of CALIFORNIA ST and POLK ST   
4   Parking_Meter        925 OFARRELL ST, SAN FRANCISCO, CA, 94109   

          Street  Supervisor District                  Neighborhood  \
0      HAIGHT ST                  5.0                Haight Ashbury   
1  SACRAMENTO ST                  3.0            Financial District   
2  CALIFORNIA ST                  2.0  Laurel Heights / Jordan Park   
3  CALIFORNIA ST                  3.0                    Polk Gulch   
4    OFARRELL ST                  6.0                    Tenderloin   

  Police District          Y           X                         Point Source  \
0            PARK  37.770008 -122.446930  (37.77000809, -122.44692993)  Phone   
1         CENTRAL  37.793781 -122.402954   (37.79378128, -122.4029541)  Phone   
2        RICHMOND  37.786232 -122.453682  (37.78623199, -122.45368195)  Phone   
3        NORTHERN  37.790577 -122.420685  (37.79057693, -122.42068481)  Phone   
4        NORTHERN  37.784645 -122.420021  (37.78464508, -122.42002106)  Phone   

  Media URL  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN

## Empty values

In [7]:
for col in df:
    print(col, df[col].isnull().sum() )

CaseID 0
Opened 0
Closed 782
Updated 0
Status 0
Status Notes 17421
Responsible Agency 0
Category 0
Request Type 0
Request Details 1
Address 0
Street 1842
Supervisor District 1842
Neighborhood 1906
Police District 1933
Y 0
X 0
Point 0
Source 0
Media URL 94990


Closed nan values are fine, because the case is still open and we will only look at the open date. Status notes and request type are also not useful. Basically we need to take care of nan values for Location, Supervisor District, Neighborhood and Police District

In [8]:
df[df['Supervisor District'].isnull()]

CaseID                  Opened                  Closed  \
36      11765218  12/03/2019 02:09:00 PM                     NaN   
57      11762576  12/03/2019 08:29:00 AM                     NaN   
208     11727994  11/23/2019 02:09:00 PM  12/03/2019 02:25:02 AM   
308     11747871  11/28/2019 03:03:00 PM                     NaN   
320     11720935  11/21/2019 07:36:17 PM  12/04/2019 09:30:47 PM   
468     11716898  11/21/2019 07:54:00 AM  11/22/2019 12:20:01 PM   
481     11674810  11/11/2019 03:34:05 PM  11/22/2019 09:30:44 PM   
638     11691512  11/15/2019 07:58:00 AM  11/18/2019 01:10:02 AM   
817     11628601  10/31/2019 06:05:05 PM  11/13/2019 09:30:32 PM   
859     11619710  10/29/2019 05:55:07 PM  11/13/2019 09:30:32 PM   
1088    11628275  10/31/2019 04:45:00 PM  11/04/2019 10:30:02 AM   
1161    11622120  10/30/2019 11:10:00 AM  11/02/2019 12:30:02 PM   
1174    11619802  10/29/2019 06:13:00 PM                     NaN   
1261    11566471  10/18/2019 04:49:00 AM  10/30/2019 09:31:03 PM   
1576    11515642  10/07/2019 09:12:05 AM  10/18/2019 09:30:38 PM   
1820    11373392  09/05/2019 12:34:00 PM  10/11/2019 07:40:01 AM   
1935    11510259  10/05/2019 05:10:00 PM  10/08/2019 07:10:01 AM   
2137    11466060  09/25/2019 05:00:00 PM  09/30/2019 07:25:02 AM   
2172    11478558  09/28/2019 03:41:00 PM  09/30/2019 10:25:02 AM   
2268    11456384  09/23/2019 05:53:00 PM  09/25/2019 10:30:02 AM   
2295     6622736  12/13/2016 08:31:41 AM  12/13/2016 09:01:33 AM   
2298     6572188  11/28/2016 12:29:35 PM  11/28/2016 12:35:27 PM   
2300     6561404  11/23/2016 04:06:36 PM  11/29/2016 09:50:03 AM   
2301     6560333  11/23/2016 12:12:26 PM  11/28/2016 02:00:04 PM   
2302     8418353  12/22/2017 02:22:00 PM  12/26/2017 01:10:04 AM   
2304     6530660  11/14/2016 03:02:42 PM  11/14/2016 09:20:05 PM   
2307     6503347  11/06/2016 11:20:05 PM  11/07/2016 07:15:05 AM   
2308     6472214  10/28/2016 04:49:34 PM  10/31/2016 08:10:06 AM   
2309     6459379  10/25/2016 11:56:54 AM  10/26/2016 02:20:04 PM   
2311     6395657  10/07/2016 05:02:57 PM  10/11/2016 11:40:03 AM   
...          ...                     ...                     ...   
103703  10562218  03/04/2019 04:44:00 PM  03/06/2019 09:10:02 AM   
103726  10582725  03/09/2019 04:39:00 PM  03/11/2019 01:30:01 AM   
103805  10587691  03/11/2019 11:24:00 AM  03/13/2019 11:00:02 AM   
103898  10976847  06/10/2019 11:33:00 AM  06/11/2019 09:03:00 AM   
104193  11071130  07/01/2019 12:18:00 PM  07/09/2019 10:30:01 AM   
104208  10973922  06/09/2019 05:12:00 PM  06/10/2019 12:20:01 PM   
104269  11092607  07/06/2019 03:27:00 PM  07/13/2019 11:45:01 AM   
104358  11088463  07/05/2019 12:59:00 PM  07/13/2019 11:40:02 AM   
104494  10937100  05/31/2019 08:05:00 PM  06/03/2019 01:39:00 PM   
105254  10781828  04/26/2019 12:57:00 PM  05/01/2019 10:23:00 AM   
105298  10906948  05/24/2019 04:59:00 PM  05/28/2019 07:52:00 AM   
105313  11037639  06/24/2019 09:16:00 AM  07/05/2019 11:00:04 AM   
105586  11049578  06/26/2019 02:20:00 PM  06/27/2019 10:00:03 AM   
105587  11028116  06/21/2019 01:10:00 PM  06/25/2019 07:50:01 AM   
105771  11270366  08/14/2019 10:46:00 AM  08/19/2019 10:40:02 AM   
105774  11381750  09/07/2019 10:21:00 AM  09/09/2019 11:54:00 AM   
105793  11156780  07/20/2019 04:49:54 PM  07/31/2019 09:30:26 PM   
105942  11152211  07/19/2019 01:47:00 PM  07/24/2019 11:50:02 AM   
106027  11350479  08/31/2019 11:16:00 AM  09/03/2019 09:10:02 AM   
106083  11106448  07/09/2019 05:51:00 PM  07/13/2019 08:40:01 AM   
106193  11142617  07/17/2019 03:18:00 PM  07/19/2019 01:50:02 AM   
106256  11351753  08/31/2019 04:46:00 PM  09/07/2019 11:30:01 AM   
106299  11217638  08/02/2019 02:04:00 PM  08/05/2019 10:35:01 AM   
106443  11324212  08/26/2019 09:53:00 AM  09/03/2019 09:00:01 AM   
106485  11233722  08/06/2019 11:51:00 AM  08/08/2019 11:10:02 AM   
106895  11344459  08/30/2019 07:34:00 AM  09/11/2019 09:30:53 PM   
107070  11314107  08/23/2019 02:10:00 PM  08/28/2019 01:35:02 AM  

We see a problem. The rows with no district are the ones with location (0,0). Since we are going to work based on locations. Drop all the locations with (0,0)

In [9]:
df = df[df.Point != '(0.0, 0.0)']
df[df['Supervisor District'].isnull()]

CaseID                  Opened                  Closed  \
57      11762576  12/03/2019 08:29:00 AM                     NaN   
320     11720935  11/21/2019 07:36:17 PM  12/04/2019 09:30:47 PM   
1174    11619802  10/29/2019 06:13:00 PM                     NaN   
1576    11515642  10/07/2019 09:12:05 AM  10/18/2019 09:30:38 PM   
2726     3960323  08/28/2014 11:08:18 AM  09/24/2014 12:23:01 PM   
2781     3858046  07/25/2014 08:34:17 AM  10/16/2014 11:28:31 AM   
35915    3093345  11/22/2013 11:06:05 AM  12/02/2013 11:27:53 AM   
36115    3352361  02/09/2014 03:31:49 PM  02/19/2014 02:46:58 PM   
36412    3073673  11/17/2013 02:02:33 PM  12/02/2013 10:40:56 AM   
37233    3209820  12/28/2013 12:11:23 PM  12/30/2013 01:29:59 PM   
37527    3344360  02/06/2014 12:35:22 PM  02/07/2014 12:57:46 PM   
37554    3243832  01/07/2014 01:40:34 PM  01/10/2014 11:14:00 AM   
37665    3321751  01/30/2014 11:15:59 AM  02/20/2014 02:21:22 PM   
37726    3144032  12/08/2013 04:48:24 PM  12/16/2013 11:03:37 AM   
37757    3082511  11/19/2013 05:21:29 PM  12/02/2013 11:15:16 AM   
37979    3380583  02/18/2014 05:16:10 PM  03/03/2014 01:49:58 PM   
38283    3287698  01/19/2014 02:43:59 PM  02/06/2014 10:51:46 AM   
38453    3271407  01/14/2014 03:54:45 PM  02/27/2014 09:38:10 AM   
39151    3348343  02/07/2014 05:17:41 PM  02/19/2014 03:06:09 PM   
39154    3333399  02/03/2014 12:52:35 PM  02/20/2014 02:11:56 PM   
39240    3192474  12/22/2013 02:55:38 PM  12/27/2013 08:43:00 AM   
39641    3166502  12/14/2013 04:46:45 PM  12/16/2013 12:19:49 PM   
39819    3327257  01/31/2014 07:43:41 PM  01/31/2014 08:21:38 PM   
40334    3373692  02/16/2014 02:12:42 PM  03/03/2014 11:06:43 AM   
40415    3266244  01/13/2014 01:40:12 PM  01/24/2014 01:29:30 PM   
40487    3333700  02/03/2014 01:39:10 PM  02/19/2014 03:30:58 PM   
40492    3171069  12/16/2013 11:55:43 AM  12/18/2013 11:50:19 AM   
41644    3348037  02/07/2014 03:43:08 PM  02/19/2014 03:07:40 PM   
41695    3145823  12/09/2013 10:18:26 AM  12/16/2013 11:05:18 AM   
41747    3162196  12/13/2013 09:15:52 AM  12/16/2013 11:53:22 AM   
...          ...                     ...                     ...   
75128    6117656  07/25/2016 10:05:06 AM  07/26/2016 10:30:06 AM   
75384    6138653  07/30/2016 04:43:52 PM  08/02/2016 03:10:05 PM   
75854    6332405  09/20/2016 03:34:47 PM  09/21/2016 11:40:05 AM   
76561    6420050  10/14/2016 10:28:37 AM  10/17/2016 12:00:25 AM   
76647    6299873  09/11/2016 09:47:11 PM  09/12/2016 05:36:19 AM   
76719    6412214  10/12/2016 12:39:07 PM  10/12/2016 03:13:20 PM   
78258    6792654  01/31/2017 04:31:36 PM  02/01/2017 10:50:05 AM   
78572    6531421  11/14/2016 05:19:19 PM  11/15/2016 12:30:04 PM   
79564    6771678  01/25/2017 05:28:01 PM  01/27/2017 12:20:05 PM   
79938    6755050  01/21/2017 10:36:23 AM  01/27/2017 08:20:12 AM   
80414    7085344  04/26/2017 02:42:18 PM  04/27/2017 09:19:48 AM   
80733    6876341  02/24/2017 03:18:21 PM  02/27/2017 10:10:04 AM   
81130    7114531  05/04/2017 02:49:20 PM  05/15/2017 09:57:13 AM   
81188    7167437  05/19/2017 02:05:57 PM  05/19/2017 03:00:05 PM   
81424    6876717  02/24/2017 04:47:43 PM  02/27/2017 03:52:44 PM   
82030    6874040  02/24/2017 07:24:45 AM  02/24/2017 09:02:49 AM   
82132    7163940  05/18/2017 02:53:01 PM  05/22/2017 02:40:05 PM   
82210    6817472  02/07/2017 04:20:18 PM  02/08/2017 01:00:04 PM   
82515    7151679  05/15/2017 02:57:42 PM  05/16/2017 02:20:06 PM   
82623    7119292  05/05/2017 06:45:13 PM  05/06/2017 08:00:06 AM   
83116    7038063  04/12/2017 12:26:29 PM  04/14/2017 01:30:06 PM   
83334    7061398  04/19/2017 01:16:50 PM  04/28/2017 08:18:04 AM   
83359    6897924  03/02/2017 10:31:36 PM  03/03/2017 05:32:08 AM   
84142    7433055  07/28/2017 12:39:06 PM  08/01/2017 10:30:06 AM   
84485    7475349  08/08/2017 09:33:22 AM  08/09/2017 11:40:03 AM   
85227    7253242  06/12/2017 06:27:31 PM  06/14/2017 07:35:03 AM   
86638    7262849  06/15/2017 09:50:43 AM  06/15/2017 12:40:03 PM  

We got rid from a lot of missing location rows, but we are still missing some supervisor districts. We can get the correct district using the location later with our class.

In [10]:
print('Nan on Neighborhood', df[df['Neighborhood'].isnull()].shape[0])
print('Nan on Police District', df[df['Police District'].isnull()].shape[0])

Nan on Neighborhood 280
Nan on Police District 307


In [11]:
df[df['Neighborhood'].isnull()]

CaseID                  Opened                  Closed  \
57      11762576  12/03/2019 08:29:00 AM                     NaN   
320     11720935  11/21/2019 07:36:17 PM  12/04/2019 09:30:47 PM   
1174    11619802  10/29/2019 06:13:00 PM                     NaN   
1576    11515642  10/07/2019 09:12:05 AM  10/18/2019 09:30:38 PM   
2726     3960323  08/28/2014 11:08:18 AM  09/24/2014 12:23:01 PM   
2781     3858046  07/25/2014 08:34:17 AM  10/16/2014 11:28:31 AM   
14523     920874  07/15/2011 12:21:54 PM  07/15/2011 04:52:22 PM   
15310     882562  05/16/2011 09:39:48 AM  05/17/2011 07:24:20 AM   
35915    3093345  11/22/2013 11:06:05 AM  12/02/2013 11:27:53 AM   
36115    3352361  02/09/2014 03:31:49 PM  02/19/2014 02:46:58 PM   
36412    3073673  11/17/2013 02:02:33 PM  12/02/2013 10:40:56 AM   
37233    3209820  12/28/2013 12:11:23 PM  12/30/2013 01:29:59 PM   
37527    3344360  02/06/2014 12:35:22 PM  02/07/2014 12:57:46 PM   
37554    3243832  01/07/2014 01:40:34 PM  01/10/2014 11:14:00 AM   
37665    3321751  01/30/2014 11:15:59 AM  02/20/2014 02:21:22 PM   
37726    3144032  12/08/2013 04:48:24 PM  12/16/2013 11:03:37 AM   
37757    3082511  11/19/2013 05:21:29 PM  12/02/2013 11:15:16 AM   
37979    3380583  02/18/2014 05:16:10 PM  03/03/2014 01:49:58 PM   
38283    3287698  01/19/2014 02:43:59 PM  02/06/2014 10:51:46 AM   
38453    3271407  01/14/2014 03:54:45 PM  02/27/2014 09:38:10 AM   
39114    3370399  02/15/2014 11:32:53 AM  02/15/2014 12:50:47 PM   
39151    3348343  02/07/2014 05:17:41 PM  02/19/2014 03:06:09 PM   
39154    3333399  02/03/2014 12:52:35 PM  02/20/2014 02:11:56 PM   
39240    3192474  12/22/2013 02:55:38 PM  12/27/2013 08:43:00 AM   
39641    3166502  12/14/2013 04:46:45 PM  12/16/2013 12:19:49 PM   
39819    3327257  01/31/2014 07:43:41 PM  01/31/2014 08:21:38 PM   
40334    3373692  02/16/2014 02:12:42 PM  03/03/2014 11:06:43 AM   
40415    3266244  01/13/2014 01:40:12 PM  01/24/2014 01:29:30 PM   
40487    3333700  02/03/2014 01:39:10 PM  02/19/2014 03:30:58 PM   
40492    3171069  12/16/2013 11:55:43 AM  12/18/2013 11:50:19 AM   
...          ...                     ...                     ...   
80414    7085344  04/26/2017 02:42:18 PM  04/27/2017 09:19:48 AM   
80416    7139721  05/11/2017 04:07:40 PM                     NaN   
80733    6876341  02/24/2017 03:18:21 PM  02/27/2017 10:10:04 AM   
81130    7114531  05/04/2017 02:49:20 PM  05/15/2017 09:57:13 AM   
81188    7167437  05/19/2017 02:05:57 PM  05/19/2017 03:00:05 PM   
81424    6876717  02/24/2017 04:47:43 PM  02/27/2017 03:52:44 PM   
82030    6874040  02/24/2017 07:24:45 AM  02/24/2017 09:02:49 AM   
82132    7163940  05/18/2017 02:53:01 PM  05/22/2017 02:40:05 PM   
82210    6817472  02/07/2017 04:20:18 PM  02/08/2017 01:00:04 PM   
82515    7151679  05/15/2017 02:57:42 PM  05/16/2017 02:20:06 PM   
82623    7119292  05/05/2017 06:45:13 PM  05/06/2017 08:00:06 AM   
83116    7038063  04/12/2017 12:26:29 PM  04/14/2017 01:30:06 PM   
83334    7061398  04/19/2017 01:16:50 PM  04/28/2017 08:18:04 AM   
83359    6897924  03/02/2017 10:31:36 PM  03/03/2017 05:32:08 AM   
83797    7494612  08/12/2017 11:46:33 AM  08/30/2017 08:06:59 AM   
84142    7433055  07/28/2017 12:39:06 PM  08/01/2017 10:30:06 AM   
84485    7475349  08/08/2017 09:33:22 AM  08/09/2017 11:40:03 AM   
85027    7435733  07/29/2017 08:32:43 AM  08/01/2017 07:20:07 AM   
85227    7253242  06/12/2017 06:27:31 PM  06/14/2017 07:35:03 AM   
85396    7477182  08/08/2017 02:10:51 PM  08/08/2017 02:49:56 PM   
85596    7489989  08/11/2017 09:58:29 AM  08/15/2017 01:22:17 AM   
86504    7291785  06/23/2017 09:02:43 AM  06/28/2017 04:25:28 PM   
86638    7262849  06/15/2017 09:50:43 AM  06/15/2017 12:40:03 PM   
86689    7343931  07/06/2017 08:42:52 PM  07/08/2017 06:01:46 AM   
86751    8112545  10/07/2017 08:22:00 AM  09/25/2018 12:00:00 AM   
88021    7569391  08/30/2017 09:31:53 AM  08/31/2017 08:30:05 AM   
88805    7609658  09/08/2017 12:07:03 PM  09/08/2017 12:08:42 PM  

There are less than 307 rows with nan out of 107762, that is 0.2%. Therefore, we will drop them without risk of losing data

In [12]:
df.dropna(subset=['Neighborhood'], inplace=True)
df.dropna(subset=['Police District'], inplace=True)
df.dropna(subset=['Supervisor District'], inplace=True)

In [13]:
print('Nan on Neighborhood', df[df['Neighborhood'].isnull()].shape[0])
print('Nan on Police District', df[df['Police District'].isnull()].shape[0])
print('Nan on Supervisor District', df[df['Supervisor District'].isnull()].shape[0])

Nan on Neighborhood 0
Nan on Police District 0
Nan on Supervisor District 0


## Datetime objects

In [14]:
df['Opened'] = pd.to_datetime(df['Opened'] )
df['Closed'] = pd.to_datetime(df['Closed'] )
df['Updated'] = pd.to_datetime(df['Updated'] )

In [15]:
df['Year'] = df['Opened'].dt.year
df['Month'] = df['Opened'].dt.month
df['Day'] = df['Opened'].dt.day

In [16]:
df.head()

CaseID              Opened              Closed             Updated  \
0  11771122 2019-12-04 16:24:00                 NaT 2019-12-04 17:00:02   
1  11770741 2019-12-04 15:14:00                 NaT 2019-12-04 17:06:56   
2  11768297 2019-12-04 09:06:00                 NaT 2019-12-04 09:45:02   
3  11769772 2019-12-04 12:47:00 2019-12-04 13:30:01 2019-12-04 13:30:01   
4  11770411 2019-12-04 14:23:00                 NaT 2019-12-04 15:00:01   

   Status                        Status Notes        Responsible Agency  \
0    Open                            accepted      DPT Meter_Bike Queue   
1    Open                                Open  DPT - Traffic Signal - G   
2    Open                            accepted      DPT Meter_Bike Queue   
3  Closed  meter shop has closed this request      DPT Meter_Bike Queue   
4    Open                            accepted      DPT Meter_Bike Queue   

           Category            Request Type Request Details  ...  \
0  Damaged Property   Damaged Parking_Meter   Parking_Meter  ...   
1  Damaged Property  Damaged Traffic_Signal  Traffic_Signal  ...   
2  Damaged Property   Damaged Parking_Meter   Parking_Meter  ...   
3  Damaged Property   Damaged Parking_Meter   Parking_Meter  ...   
4  Damaged Property   Damaged Parking_Meter   Parking_Meter  ...   

                   Neighborhood Police District          Y           X  \
0                Haight Ashbury            PARK  37.770008 -122.446930   
1            Financial District         CENTRAL  37.793781 -122.402954   
2  Laurel Heights / Jordan Park        RICHMOND  37.786232 -122.453682   
3                    Polk Gulch        NORTHERN  37.790577 -122.420685   
4                    Tenderloin        NORTHERN  37.784645 -122.420021   

                          Point  Source  Media URL  Year Month Day  
0  (37.77000809, -122.44692993)   Phone        NaN  2019    12   4  
1   (37.79378128, -122.4029541)   Phone        NaN  2019    12   4  
2  (37.78623199, -122.45368195)   Phone        NaN  2019    12   4  
3  (37.79057693, -122.42068481)   Phone        NaN  2019    12   4  
4  (37.78464508, -122.42002106)   Phone        NaN  2019    12   4  

[5 rows x 23 columns]

## Readability

In [17]:
df['Police District'] = df['Police District'].astype(str).str.lower() 
df['Supervisor District'] = df['Supervisor District'].astype(str).str.lower() 
df['Neighborhood'] = df['Neighborhood'].astype(str).str.lower() 
df['Address'] = df['Address'].astype(str).str.lower() 
df.head()

CaseID              Opened              Closed             Updated  \
0  11771122 2019-12-04 16:24:00                 NaT 2019-12-04 17:00:02   
1  11770741 2019-12-04 15:14:00                 NaT 2019-12-04 17:06:56   
2  11768297 2019-12-04 09:06:00                 NaT 2019-12-04 09:45:02   
3  11769772 2019-12-04 12:47:00 2019-12-04 13:30:01 2019-12-04 13:30:01   
4  11770411 2019-12-04 14:23:00                 NaT 2019-12-04 15:00:01   

   Status                        Status Notes        Responsible Agency  \
0    Open                            accepted      DPT Meter_Bike Queue   
1    Open                                Open  DPT - Traffic Signal - G   
2    Open                            accepted      DPT Meter_Bike Queue   
3  Closed  meter shop has closed this request      DPT Meter_Bike Queue   
4    Open                            accepted      DPT Meter_Bike Queue   

           Category            Request Type Request Details  ...  \
0  Damaged Property   Damaged Parking_Meter   Parking_Meter  ...   
1  Damaged Property  Damaged Traffic_Signal  Traffic_Signal  ...   
2  Damaged Property   Damaged Parking_Meter   Parking_Meter  ...   
3  Damaged Property   Damaged Parking_Meter   Parking_Meter  ...   
4  Damaged Property   Damaged Parking_Meter   Parking_Meter  ...   

                   Neighborhood Police District          Y           X  \
0                haight ashbury            park  37.770008 -122.446930   
1            financial district         central  37.793781 -122.402954   
2  laurel heights / jordan park        richmond  37.786232 -122.453682   
3                    polk gulch        northern  37.790577 -122.420685   
4                    tenderloin        northern  37.784645 -122.420021   

                          Point  Source  Media URL  Year Month Day  
0  (37.77000809, -122.44692993)   Phone        NaN  2019    12   4  
1   (37.79378128, -122.4029541)   Phone        NaN  2019    12   4  
2  (37.78623199, -122.45368195)   Phone        NaN  2019    12   4  
3  (37.79057693, -122.42068481)   Phone        NaN  2019    12   4  
4  (37.78464508, -122.42002106)   Phone        NaN  2019    12   4  

[5 rows x 23 columns]

In [18]:
print("Min date: ",min(df['Opened']))
print("Max date: ",max(df['Opened']))

Min date:  2008-07-01 06:29:45
Max date:  2019-12-05 00:55:00


In [19]:
df.to_csv('../../data/processed/damagedProperty.csv')